File: ios/MapboxNavigationFreeDriveManager.m
#import "React/RCTViewManager.h"

@interface RCT_EXTERN_MODULE(MapboxNavigationFreeDriveManager, RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(onLocationChange, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onTrackingStateChange, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onRouteChange, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onError, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onManeuverSizeChange, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(showSpeedLimit, BOOL)
RCT_EXPORT_VIEW_PROPERTY(speedLimitAnchor, NSArray)
RCT_EXPORT_VIEW_PROPERTY(maneuverAnchor, NSArray)
RCT_EXPORT_VIEW_PROPERTY(followZoomLevel, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(userPuckImage, UIImage)
RCT_EXPORT_VIEW_PROPERTY(userPuckScale, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(originImage, UIImage)
RCT_EXPORT_VIEW_PROPERTY(destinationImage, UIImage)
RCT_EXPORT_VIEW_PROPERTY(mapPadding, NSArray)
RCT_EXPORT_VIEW_PROPERTY(logoVisible, NSArray)
RCT_EXPORT_VIEW_PROPERTY(logoPadding, NSArray)
RCT_EXPORT_VIEW_PROPERTY(attributionVisible, NSArray)
RCT_EXPORT_VIEW_PROPERTY(attributionPadding, NSArray)
RCT_EXPORT_VIEW_PROPERTY(routeColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(routeCasingColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(routeClosureColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(alternateRouteColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(alternateRouteCasingColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(traversedRouteColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(traversedRouteCasingColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(trafficUnknownColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(trafficLowColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(trafficModerateColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(trafficHeavyColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(trafficSevereColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(restrictedRoadColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(waypointColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(waypointRadius, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(waypointOpacity, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(waypointStrokeWidth, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(waypointStrokeOpacity, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(waypointStrokeColor, NSString)
RCT_EXPORT_VIEW_PROPERTY(mute, BOOL)
RCT_EXPORT_VIEW_PROPERTY(darkMode, BOOL)
RCT_EXPORT_VIEW_PROPERTY(debug, BOOL)

RCT_EXTERN_METHOD(
  showRouteViaManager: (nonnull NSNumber *)node 
  origin: (NSArray *)origin 
  destination: (NSArray *)destination 
  waypoints: (NSArray *)waypoints 
  styles: (NSDictionaryArray *)styles 
  legIndex: (nonnull NSNumber *)legIndex 
  cameraType: (NSString *)cameraType 
  padding: (NSArray *)padding
)

RCT_EXTERN_METHOD(
  clearRouteViaManager: (nonnull NSNumber *)node
)

RCT_EXTERN_METHOD(
  followViaManager: (nonnull NSNumber *)node
)

RCT_EXTERN_METHOD(
  moveToOverviewViaManager: (nonnull NSNumber *)node
  padding: (NSArray *)padding
)

RCT_EXTERN_METHOD(
  fitCameraViaManager: (nonnull NSNumber *)node
  padding: (NSArray *)padding
)

RCT_EXTERN_METHOD(
  startNavigationViaManager: (nonnull NSNumber *)node
  origin: (NSArray *)origin 
  destination: (NSArray *)destination 
  waypoints: (NSArray *)waypoints 
  styles: (NSDictionaryArray *)styles 
  legIndex: (nonnull NSNumber *)legIndex 
  cameraType: (NSString *)cameraType 
  padding: (NSArray *)padding
)

RCT_EXTERN_METHOD(
  pauseNavigationViaManager: (nonnull NSNumber *)node
)

RCT_EXTERN_METHOD(
  stopNavigationViaManager: (nonnull NSNumber *)node
)

@end

File: ios/MapboxNavigationFreeDriveManager.swift
@objc(MapboxNavigationFreeDriveManager)
class MapboxNavigationFreeDriveManager: RCTViewManager {
  override func view() -> UIView! {
    return MapboxNavigationFreeDriveView();
  }

  override static func requiresMainQueueSetup() -> Bool {
    return true
  }

  @objc func showRouteViaManager(_ node: NSNumber, origin: [NSNumber], destination: [NSNumber], waypoints: [[NSNumber]], styles: [NSDictionary], legIndex: NSNumber, cameraType: NSString, padding: [NSNumber]) {
    DispatchQueue.main.async {
      let mapboxNavigationFreeDriveView = self.bridge.uiManager.view(forReactTag: node) as! MapboxNavigationFreeDriveView
      
      mapboxNavigationFreeDriveView.showRoute(origin: origin, destination: destination, waypoints: waypoints, styles: styles, legIndex: legIndex, cameraType: cameraType, padding: padding)
    }
  }

  @objc func clearRouteViaManager(_ node: NSNumber) {
    DispatchQueue.main.async {
      let mapboxNavigationFreeDriveView = self.bridge.uiManager.view(forReactTag: node) as! MapboxNavigationFreeDriveView
      
      mapboxNavigationFreeDriveView.clearRoute()
    }
  }

  @objc func followViaManager(_ node: NSNumber, padding: [NSNumber]) {
    DispatchQueue.main.async {
      let mapboxNavigationFreeDriveView = self.bridge.uiManager.view(forReactTag: node) as! MapboxNavigationFreeDriveView
      
      mapboxNavigationFreeDriveView.follow(padding: padding)
    }
  }

  @objc func moveToOverviewViaManager(_ node: NSNumber, padding: [NSNumber]) {
    DispatchQueue.main.async {
      let mapboxNavigationFreeDriveView = self.bridge.uiManager.view(forReactTag: node) as! MapboxNavigationFreeDriveView
      
      mapboxNavigationFreeDriveView.moveToOverview(padding: padding)
    }
  }

  @objc func fitCameraViaManager(_ node: NSNumber, padding: [NSNumber]) {
    DispatchQueue.main.async {
      let mapboxNavigationFreeDriveView = self.bridge.uiManager.view(forReactTag: node) as! MapboxNavigationFreeDriveView
      
      mapboxNavigationFreeDriveView.fitCamera(padding: padding)
    }
  }

  @objc func startNavigationViaManager(_ node: NSNumber, origin: [NSNumber], destination: [NSNumber], waypoints: [[NSNumber]], styles: [NSDictionary], legIndex: NSNumber, cameraType: NSString, padding: [NSNumber]) {
    DispatchQueue.main.async {
      let mapboxNavigationFreeDriveView = self.bridge.uiManager.view(forReactTag: node) as! MapboxNavigationFreeDriveView
      
      mapboxNavigationFreeDriveView.startNavigation(
        origin: origin, destination: destination, waypoints: waypoints, styles: styles, legIndex: legIndex, cameraType: cameraType, padding: padding)
    }
  }

  @objc func pauseNavigationViaManager(_ node: NSNumber) {
    DispatchQueue.main.async {
      let mapboxNavigationFreeDriveView = self.bridge.uiManager.view(forReactTag: node) as! MapboxNavigationFreeDriveView
      
      mapboxNavigationFreeDriveView.pauseNavigation()
    }
  }

  @objc func stopNavigationViaManager(_ node: NSNumber) {
    DispatchQueue.main.async {
      let mapboxNavigationFreeDriveView = self.bridge.uiManager.view(forReactTag: node) as! MapboxNavigationFreeDriveView
      
      mapboxNavigationFreeDriveView.stopNavigation()
    }
  }
}
